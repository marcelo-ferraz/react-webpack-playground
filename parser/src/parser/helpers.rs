use std::collections::HashMap;
use wasm_bindgen::{JsValue, prelude::*};
use lazy_static::lazy_static;
use regex::Regex;

pub fn find_imports(js_value: &str) -> impl Iterator<Item = &str> {
    lazy_static! {
        static ref IMPORTS_REGEX: Regex = Regex::new(
        r#"import[ \s\w\t\r\n\f,_\{\}]+from[ \s]+("|')+(?P<import_id>.+)("|')[ \s]*;"#
            // actual regex to be used, after converted 
            // r#"__customRequire\(__webpack_require__,\s*\\?("|')(?<import_id>\.\.?\/[\w.\s/-]+)\\?("|')"#
        ).unwrap();
    }

    IMPORTS_REGEX
        .captures_iter(js_value)
        .map(|c| c.name("import_id"))
        .flatten()
        .map(|m| m.as_str())
}

pub fn compare_key_to_path(key: &str, path: &str) -> bool {
    key == path || 
    key == format!("{}.js", path) || 
    key == format!("{}/index.js", path)
}

pub fn find_by_path<'v>(entries: &'v HashMap<String, String>, path: &'v str)  -> Option<&'v String> {
    entries
        .iter()
        .find(| item | compare_key_to_path(item.0, path))
        .map(|item| item.1)
}

// it will call a function to convert that uses babel standalone
pub fn transform(value: &str) -> String {
    format!("#{}#", value)
}

pub fn get_ext(path: &str) -> Option<&str> {
    lazy_static!{ 
        static ref EXT_REG: Regex = Regex::new(r#"\.\w+$"#).unwrap();
    }

    EXT_REG.captures(path)
        .and_then(|cap| cap.get(0))
        .and_then(| m | Some(m.as_str()))
}

#[wasm_bindgen(raw_module = "../src/transpiling/transformers.js", )]
extern "C" {
    #[wasm_bindgen(js_name = "reactJsTransform" )]
    pub fn react_js_transform(rawCode: &str) -> String;
}
