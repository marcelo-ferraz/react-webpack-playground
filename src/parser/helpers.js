export const isItMeaningful = (code) => {
    return String.prototype.replace.call(code || '', /(\r\n|\n|\r)/gm, '');
};

export function tryResolve(modules, path) {
    if (modules[path]) {
        return path;
    }

    const isANodeModule = !path.startsWith('.') && !path.startsWith('@') && !path.indexOf(':') >= 0;

    const resolvedUrl = isANodeModule ? `./node_modules/${path}` : path;

    if (modules[resolvedUrl]) {
        return resolvedUrl;
    }

    if (modules[`${resolvedUrl}.js`]) {
        return `${resolvedUrl}.js`;
    }

    if (modules[`${resolvedUrl}/index.js`]) {
        return `${resolvedUrl}/index.js`;
    }
}
