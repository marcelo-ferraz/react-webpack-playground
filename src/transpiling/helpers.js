export const byEqual =
    (path) =>
    ([key]) =>
        key === path || key === `${path}.js` || key === `${path}/index.js`;

export const pathFinder = (obj) => {
    const find = Array.prototype.find.bind(Object.entries(obj));
    return (path) => find(byEqual(path)) || [];
};

export const findAllImports = (code, whenFound) => {
    const findImports =
        /__customRequire\(__webpack_require__,\s*\\?("|')(?<importId>\.\.?\/[\w.\s/-]+)\\?("|')/gm;
    let match;

    do {
        match = findImports.exec(code);

        if (match && match.groups) {
            const { importId } = match.groups;
            whenFound(importId);
        }
    } while (match);
};

export const isItMeaningful = (code) => {
    return String.prototype.replace.call(code || '', /(\r\n|\n|\r)/gm, '');
};

export function tryResolve(modules, path) {
    if (modules[path]) {
        return path;
    }

    const isANodeModule = !path.startsWith('.') && !path.startsWith('@') && !path.indexOf(':') >= 0;

    const resolvedUrl = isANodeModule ? `./node_modules/${path}` : path;

    if (modules[resolvedUrl]) {
        return resolvedUrl;
    }

    if (modules[`${resolvedUrl}.js`]) {
        return `${resolvedUrl}.js`;
    }

    if (modules[`${resolvedUrl}/index.js`]) {
        return `${resolvedUrl}/index.js`;
    }
}
