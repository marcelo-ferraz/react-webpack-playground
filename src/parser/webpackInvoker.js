import * as Babel from '@babel/standalone';
import { find } from 'lodash';
import { resolve as pathResolve, dirname, basename, extname } from 'path';
import { js, json, jsx } from '../fileExtensions';
import { isItMeaningful } from './helpers';
import { resolve } from './helpers';

const findByPath = (obj, path) =>
    find(
        Object.entries(obj),
        ([key]) => key === path || key === `${path}.js` || key === `${path}/index.js`,
    ) || [];

const emptyFunc = function () {
    // eslint-disable-next-line no-useless-return
    return;
};

function reduceAllImports(code, whenFound) {
    const findImports = /__customRequire\(\\?("|')(?<importId>\.\.?\/[\w-\.\s\/]+)\\?("|')/gm;

    let match;
    do {
        match = findImports.exec(code);

        if (match && match.groups) {
            const { importId } = match.groups;
            whenFound(importId);
        }
    } while (match);
}

export async function render(entries, entryPath = null) {
    let allDependencies = {};
    let loadingDependencies = {};

    // eslint-disable-next-line no-unused-vars, no-underscore-dangle
    const __customRequire = (path) => {
        debugger;
        const [keyFound] = findByPath(allDependencies, path);

        if (!keyFound) {
            /* eslint-disable-next-line no-undef */
            return __webpack_require__(resolve(path));
        }

        const { exports } = jsInvoke(allDependencies[keyFound]);

        return exports;
    };

    const parse = (rawCode) => {
        const { code } = Babel.transform(rawCode, { presets: ['es2015', 'react'] });
        return code.replace(/require\(/gm, '__customRequire(');
    };

    const renderImpl = async (path) => {
        const [, rawCode] = findByPath(entries, path);
        const unit = { i: `dynamic:${path}`, l: false, exports: {} };
        if (!isItMeaningful(rawCode)) {
            return { func: emptyFunc, unit };
        }
        const type = extname(path);

        switch (type) {
            case '':
            case js:
            case jsx: {
                const code = parse(rawCode);

                reduceAllImports(code, (id) => {
                    // debugger;
                    if (loadingDependencies[id]) {
                        return;
                    }
                    loadingDependencies = { ...loadingDependencies, [id]: renderImpl(id) };
                });

                // eslint-disable-next-line no-eval
                const func = eval(`(function(module, exports, __webpack_require__) { ${code} })`);
                return [path, { func, unit }];
            }
            case json: {
                // debugger;
                unit.exports = JSON.parse(rawCode);
                return [path, { unit }];
            }
            default:
                throw new TypeError(`${type} is not supported`);
        }
    };

    // render the tree starting from the root (main.js)
    const [, app] = await renderImpl(entryPath || './app.js');
    debugger;
    // wait for all the imports to finish parsing
    allDependencies = (await Promise.all(Object.values(loadingDependencies))).reduce(
        (imports, [id, dependency]) => {
            return { ...imports, [id]: dependency };
        },
        {},
    );

    // await Object
    //     // from all loading dependencies
    //     .entries(loadingDependencies)
    //     // wait for every single one
    //     .reduce((previous, [id, loadingDependency]) => {
    //         return previous.then(([resolvedId, dependency]) => {
    //             allDependencies = resolvedId && { ...allDependencies, [resolvedId]: dependency };
    //             return [id, loadingDependency];
    //         });
    //     }, Promise.resolve([]));
    debugger;
    return app;
}

export function jsInvoke({ func, unit }) {
    if (func) {
        func.call(
            unit.exports,
            unit,
            unit.exports, // eslint-disable-next-line no-undef
            __webpack_require__,
        );
    }
    return unit;
}
