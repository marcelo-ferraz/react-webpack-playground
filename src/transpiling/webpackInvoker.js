import { js, json, jsx } from '../fileExtensions';
import { pathFinder } from './helpers';
import { tryResolve } from './helpers';
import MissingModuleError from './errors/MissingModuleError';
import parseNShake from './parseNShake';

// eslint-disable-next-line no-undef
const customRequireImpl = function (use = __webpack_require__, dependencies, path) {
    let fullPath = tryResolve(dependencies, path);

    if (fullPath) {
        const dependency = dependencies[fullPath];

        if (dependency.__esModule) {
            return dependency;
        }

        const invoked = jsInvoke(dependency);
        return invoked.exports || invoked;
    }

    fullPath = tryResolve(use.m, path);
    if (fullPath) {
        return use(fullPath);
    }

    try {
        return use(path);
    } catch (error) {
        throw new MissingModuleError(path, error);
    }
};

const loadAllExtraDependencies = async ({ getExtraDependencies }) => {
    if (!getExtraDependencies) {
        return {};
    }

    const dependencies = getExtraDependencies();
    const dependencyValues = Promise.all(Object.values(dependencies));
    const dependencyKeys = Object.keys(dependencies);

    return (await dependencyValues).reduce((deps, depValue, i) => {
        return { ...deps, [dependencyKeys[i]]: depValue };
    }, {});
};

let parserWorker;
const defaultEntryPath = './App';

const getParserWorker = () => {
    if (!parserWorker) {
        if (!window.Worker) {
            throw new Error('The browser doesnt support web workers');
        }

        parserWorker = new Worker(new URL('../workers/parserWorker', import.meta.url));
    }

    return parserWorker;
};

async function renderHere(strategy, entryPath, finder) {
    const jsFinder = finder || pathFinder(strategy.entries);

    const es5Entries = parseNShake(jsFinder, entryPath || defaultEntryPath);

    return await renderImpl(es5Entries, entryPath, strategy);
}

async function renderElsewhere(strategy, entryPath = defaultEntryPath) {
    const worker = getParserWorker();
    worker.postMessage([strategy.entries, entryPath]);

    return new Promise((resolve, reject) => {
        worker.onmessage = async ({ data: entries }) => {
            try {
                resolve(await renderImpl(entries, entryPath, strategy));
            } catch (e) {
                reject(e);
            }
        };
    });
}

function jsInvoke(context) {
    if (context.func) {
        if (context.events && context.events.beforeInvoke) {
            context.events.beforeInvoke(context);
        }

        context.func(context.unit, context.unit.exports, __webpack_require__);

        if (context.events && context.events.afterInvoke) {
            context.events.afterInvoke(context);
        }
    }
    return context.unit;
}

async function renderImpl(entries, entryPath, strategy) {
    let dynamicContext = null;
    let dependencies = {};

    // this custom require is the entry point to the module system compiled in dev mode
    // eslint-disable-next-line no-unused-vars, no-underscore-dangle
    function __customRequire(req, path) {
        return customRequireImpl(req, dependencies, path);
    }

    let extraDependencies = (await loadAllExtraDependencies(strategy)) || {};

    if (strategy.beforeRender) {
        extraDependencies = strategy.beforeRender(strategy, extraDependencies);
    }

    dependencies = { ...dependencies, ...extraDependencies };

    for (const [path, entry] of Object.entries(entries)) {
        const { ext, code } = entry;
        const unit = { i: `dynamic:${path}`, l: false, exports: {} };

        if (!code) {
            continue;
        }

        switch (ext) {
            case '':
            case js:
            case jsx: {
                // eslint-disable-next-line no-eval
                const func = eval(code);

                const ctx = {
                    func,
                    unit,
                    customRequire: __customRequire,
                };

                if (path === entryPath) {
                    dynamicContext = ctx;
                }

                const id = path.endsWith(ext) ? path.substring(0, path.length - ext.length) : path;

                dependencies[id] = ctx;
                break;
            }
            case json: {
                unit.exports = code;
                dependencies[path] = { unit };
                break;
            }
        }
    }

    if (dynamicContext && dynamicContext.func) {
        dynamicContext.customRequire = (path) =>
            customRequireImpl(__webpack_require__, dependencies, path);
        dynamicContext.addDependencies = (key, value) => {
            dependencies = { ...dependencies, [key]: value };
        };

        dynamicContext.events = {
            beforeRender: strategy.beforeRender,
            beforeInvoke: strategy.beforeInvoke,
            afterInvoke: strategy.afterInvoke,
        };

        return dynamicContext;
    }
}

export { renderHere, renderElsewhere, jsInvoke, defaultEntryPath };
