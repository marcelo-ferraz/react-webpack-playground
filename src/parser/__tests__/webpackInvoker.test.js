import React from 'react';
import { render as testRender } from '@testing-library/react';
import cases from 'jest-in-case';
import { jsInvoke, render } from '../webpackInvoker';
import { resolve, isItMeaningful } from '../helpers';
import sumAsDefaultOnlyMock from '../__mocks__/sumAsDefaultOnly.mock';
import sumAsDefaultAndSubtractionMock from '../__mocks__/sumAsDefaultAndSubtraction.mock';
import sumAndSubtractionNoDefaultMock from '../__mocks__/sumAndSubtractionNoDefault';
import sumAndSubtractionFromOtherModulesMock from '../__mocks__/sumAndSubtractionFromOtherModules.mock ';

global.__webpack_require__ = function (id) {};
global.__webpack_require__.m = [{ './node_modules/react': {} }];

jest.mock('../helpers', () => ({
    resolve: (p) => p,
    isItMeaningful: () => true,
}));

describe('webpackInvoker', () => {
    let left;
    let right;
    let expectedSumResult;
    let expectedSubtractionResult;

    beforeEach(() => {
        left = Math.random();
        right = Math.random();
        expectedSumResult = left + right;
        expectedSubtractionResult = left - right;
    });

    it('should invoke the module with sum function as default', async () => {
        const { exports } = jsInvoke(await render(sumAsDefaultOnlyMock));
        const { default: sum } = exports;

        const result = sum(left, right);

        expect(result).toEqual(expectedSumResult);
    });

    it('should invoke the module with sum function as default and subtract as a member', async () => {
        const { exports } = jsInvoke(await render(sumAsDefaultAndSubtractionMock));
        const { default: sum, subtract } = exports;

        const sumResult = sum(left, right);
        const subtractionResult = subtract(left, right);

        expect(sumResult).toEqual(expectedSumResult);
        expect(subtractionResult).toEqual(expectedSubtractionResult);
    });

    it('should invoke the module with sum and subtract as a members and no default', async () => {
        const { exports } = jsInvoke(await render(sumAndSubtractionNoDefaultMock));
        const { sum, subtract } = exports;

        const sumResult = sum(left, right);
        const subtractionResult = subtract(left, right);

        expect(sumResult).toEqual(expectedSumResult);
        expect(subtractionResult).toEqual(expectedSubtractionResult);
    });

    it('should invoke the module with sum and subtract as a members and they are exported from other modules', async () => {
        const { exports } = jsInvoke(await render(sumAndSubtractionFromOtherModulesMock));
        const { sum, subtract } = exports;

        const sumResult = sum(left, right);
        const subtractionResult = subtract(left, right);

        expect(sumResult).toEqual(expectedSumResult);
        expect(subtractionResult).toEqual(expectedSubtractionResult);
    });
});
