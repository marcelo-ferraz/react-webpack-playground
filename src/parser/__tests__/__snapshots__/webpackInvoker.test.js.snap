// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webpack invoker happy path for react two components and json 1`] = `
<div>
  <div
    class="App"
  >
    <h1>
      Hello CodeSandbox
    </h1>
    <h2>
      Start editing to see some magic happen!
    </h2>
    <div>
      here, some more magic,
    </div>
    <div>
      and here some json: 
    </div>
  </div>
</div>
`;
